src/Cryptol/Symbolic/SBV.hs

```diff
@@ -114,135 +118,217 @@ proverError :: String -> M.ModuleCmd (Maybe String, ProverResult)
+  (firstProver, timeElapsed, res) <- M.io (callSolvers provers e)
+  when pcVerbose $
+    lPutStrLn $ "Got result from " ++ show firstProver ++
+                                      ", time: " ++ showTDiff timeElapsed
+  return (Just (show firstProver), processResult res)
+
+-- | Select the appropriate solver or solvers from the given prover command,
+--   and invoke those solvers on the given symbolic value.
+runProver :: ProverCommand -> SBV.Symbolic SBV.SVal -> M.ModuleT IO (Maybe String, [SBV.SMTResult])
+runProver pc@ProverCommand{..} x =
+  do let mSatNum = case pcQueryType of
+                     SatQuery (SomeSat n) -> Just n
+                     SatQuery AllSat -> Nothing
+                     ProveQuery -> Nothing
+
+     provers <-
+       case pcProverName of
+         "any" -> M.io SBV.sbvAvailableSolvers
+         "sbv-any" -> M.io SBV.sbvAvailableSolvers
+         _ -> return [(lookupProver pcProverName)
+                        { SBV.transcript = pcSmtFile
+                        , SBV.allSatMaxModelCount = mSatNum
+                        }]
+
+     let provers' = [ p { SBV.timing = SaveTiming pcProverStats
+                        , SBV.verbose = pcVerbose
+                        , SBV.validateModel = pcValidate
+                        }
+                    | p <- provers
+                    ]
+
+     case pcQueryType of
+        ProveQuery -> runMultiProvers pc provers' SBV.proveWithAny thmSMTResults x
+        SatQuery sn -> case sn of
+          SomeSat 1 -> runMultiProvers pc provers' SBV.satWithAny satSMTResults x
+          _         -> runSingleProver pc provers' SBV.allSatWith allSatSMTResults x
+
+
+-- | Prepare a symbolic query by symbolically simulating the expression found in
+--   the @ProverQuery@.  The result will either be an error or a list of the types
+--   of the symbolic inputs and the symbolic value to supply to the solver.
+--
+--   Note that the difference between sat and prove queries is reflected later
+--   in `runProver` where we call different SBV methods depending on the mode,
+--   so we do _not_ negate the goal here.  Moreover, assumptions are added
+--   using conjunction for sat queries and implication for prove queries.
+--
+--   For safety properties, we want to add them as an additional goal
+--   when we do prove queries, and an additional assumption when we do
+--   sat queries.  In both cases, the safety property is combined with
+--   the main goal via a conjunction.
+prepareQuery ::
+  Eval.EvalOpts ->
+  M.ModuleEnv ->
+  ProverCommand ->
+  IO (Either String ([FinType], SBV.Symbolic SBV.SVal))
+prepareQuery evo modEnv ProverCommand{..} =
+  do let extDgs = M.allDeclGroups modEnv ++ pcExtraDecls
+     let tyFn = case pcQueryType of
+           SatQuery _ -> existsFinType
+           ProveQuery -> forallFinType
+     let addAsm = case pcQueryType of
+           ProveQuery -> \x y -> SBV.svOr (SBV.svNot x) y
+           SatQuery _ -> \x y -> SBV.svAnd x y
```

Comment:
It would probably be useful to explain what `addAsm` is about. Looking at the `foldr addAsm (SBV.svAnd safety b) asms` bit below, I was erroneously thinking that `addAsm` had something to do with safety predicates... but it doesn't.
Finally I remembered that `addAsm` is actually for adding range assumptions like `0 <= x < n` for quantified variables `x : Z n`. And I'm the one who wrote that code. If an explanation would help me, it would certainly be helpful for everyone else.